invoiceregistry.go

```diff
@@ -187,36 +307,121 @@ func (i *invoiceRegistry) notifyClients(invoice *channeldb.Invoice, settle bool)
 // invoice, a copy of the invoice will be sent over the SettledInvoices
 // channel.
 type invoiceSubscription struct {
-	NewInvoices     chan *channeldb.Invoice
+	// NewInvoices is a channel that we'll use to send all newly created
+	// invoices with an invoice index greater than the specified
+	// StartingInvoiceIndex field.
+	NewInvoices chan *channeldb.Invoice
+
+	// SettledInvoices is a channel that we'll use to send all setted
+	// invoices with an invoices index greater than the specified
+	// StartingInvoiceIndex field.
 	SettledInvoices chan *channeldb.Invoice
 
+	// addIndex is the highest add index the caller knows of. We'll use
+	// this information to send out an event backlog to the notifications
+	// subscriber. Any new add events with an index greater than this will
+	// be dispatched before any new notifications are sent out.
+	addIndex uint64
+
+	// settleIndex is the highest settle index the caller knows of. We'll
+	// use this information to send out an event backlog to the
+	// notifications subscriber. Any new settle events with an index
+	// greater than this will be dispatched before any new notifications
+	// are sent out.
+	settleIndex uint64
+
+	ntfnQueue *chainntnfs.ConcurrentQueue
+
+	id uint32
+
 	inv *invoiceRegistry
-	id  uint32
+
+	cancelChan chan struct{}
+
+	wg sync.WaitGroup
 }
 
 // Cancel unregisters the invoiceSubscription, freeing any previously allocated
 // resources.
 func (i *invoiceSubscription) Cancel() {
-	i.inv.clientMtx.Lock()
-	delete(i.inv.notificationClients, i.id)
-	i.inv.clientMtx.Unlock()
+	select {
+	case i.inv.subscriptionCancels <- i.id:
+	case <-i.inv.quit:
+	}
+
+	i.ntfnQueue.Stop()
+	close(i.cancelChan)
```

Comment:
multiple calls to `Cancel` would cause this to panic

Reply:
Fixed.